### PR TITLE
[6/n][Feed Storage] Implement receive_crop

### PR DESCRIPTION
**Pull request by PooyaHekmati**
**IMPORTANT NOTE**: This PR actually modifies less than 300 lines of code, the numbers are inflated due to sphinx-build action, which I disabled after making this PR. Simply ignore the files that are the `docs` directory.
This PR implements `receive_crop`, ensuring that only acceptable crops can be stored in storage. The compatibility check is done via the following rules:

- Small grain goes to hay, grain, silage, and baleage
- Alfalfa goes to hay, silage, and balayage
- Grass goes to hay, silage, and balayage
- Corn goes to grain and silage
- Soy only goes to grain

While [the design doc](https://docs.google.com/document/d/1C86lch2dkYBERR-qLDmE7L7n1gYQjETrIKUyrQpWrPs/edit?usp=sharing) states these rules should be implemented in the `FeedManager` class, here I decided to leverage the existing inheritance structure for a cleaner implementation.

The PR also moves `CropCategory` and `CropType` to `enums.py` for better organization. And implements `stored_mass`.

### CODE REVIEW

File: RUFAS/routines/feed_storage/enums.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+class CropCategory(Enum):
+    """
+    Enum for Crop Categories.
+
+    Attributes:
+        SMALL_GRAIN: Represents small grain crops.
+        CORN: Represents corn crops.
+        SOY: Represents soy crops.
+        GRASS: Represents grass crops.
+        ALFALFA: Represents alfalfa crops.
```
Comment (Isgotkowitz) [inline RUFAS/routines/feed_storage/enums.py]:
```suggestion
    Attributes
    ----------
    SMALL_GRAIN : str
        Represents small grain crops.
    CORN : str
        Represents corn crops.
    SOY : str
        Represents soy crops.
    GRASS : str
        Represents grass crops.
    ALFALFA : str
        Represents alfalfa crops.
```

File: RUFAS/routines/feed_storage/enums.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        SILAGE: Type of Corn.
+        WHOLE_PLANT: Type of Corn.
+        GRAIN: Type of Corn or Soy.
+        FORAGE: Type of Soy.
+        ALFALFA: Type of Alfalfa.
+        RYEGRASS: Type of Grass.
+        ORCHARDGRASS: Type of Grass.
+        FINE_FESCUE: Type of Grass.
+        TALL_FESCUE: Type of Grass.
+        MEADOW_FESCUE: Type of Grass.
```
Comment (Isgotkowitz) [inline RUFAS/routines/feed_storage/enums.py]:
```suggestion
    Attributes
    ----------
    WHEAT : str
        Type of Small Grain.
    RYE : str
        Type of Small Grain.
    OAT : str
        Type of Small Grain.
    RICE : str
        Type of Small Grain.
    HIGH_MOISTURE : str
        Type of Corn.
    SILAGE : str
        Type of Corn.
    WHOLE_PLANT : str
        Type of Corn.
    GRAIN : str
        Type of Corn or Soy.
    FORAGE : str
        Type of Soy.
    ALFALFA : str
        Type of Alfalfa.
    RYEGRASS : str
        Type of Grass.
    ORCHARDGRASS : str
        Type of Grass.
    FINE_FESCUE : str
        Type of Grass.
    TALL_FESCUE : str
        Type of Grass.
    MEADOW_FESCUE : str
        Type of Grass.
```

File: RUFAS/routines/feed_storage/enums.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    Attributes:
+        WHEAT: Type of Small Grain.
+        RYE: Type of Small Grain.
+        OAT: Type of Small Grain.
+        RICE: Type of Small Grain.
+        HIGH_MOISTURE: Type of Corn.
+        SILAGE: Type of Corn.
+        WHOLE_PLANT: Type of Corn.
+        GRAIN: Type of Corn or Soy.
+        FORAGE: Type of Soy.
```
Comment (Isgotkowitz) [inline RUFAS/routines/feed_storage/enums.py]: If soy is harvested as a "forage", does it still get treated as a grain? I looked at the design doc on page 11, at the crop to storage mapping, and soy only has a mark for grain.

File: RUFAS/routines/feed_storage/baleage.py [hunk 4]
```diff
@@ -16,8 +17,13 @@ class Baleage(Storage):
...
 
-    def __init__(self):
-        super().__init__()
+    def __init__(self, capacity: float = float("inf")):
+        super().__init__(capacity)
+        self.acceptable_crops = [
+            CropCategory.ALFALFA,
+            CropCategory.GRASS,
+            CropCategory.SMALL_GRAIN,
+        ]
```
Comment (Isgotkowitz) [inline RUFAS/routines/feed_storage/baleage.py:26]: Can you please add this to the attributes section?

File: RUFAS/routines/feed_storage/hay.py [hunk 5]
```diff
@@ -18,6 +19,14 @@ class Hay(Storage):
         Calculates the protein loss in the hay.
     """
 
+    def __init__(self, capacity: float = float("inf")):
+        super().__init__(capacity)
+        self.acceptable_crops = [
+            CropCategory.ALFALFA,
+            CropCategory.GRASS,
+            CropCategory.SMALL_GRAIN,
+        ]
```
Comment (Isgotkowitz) [inline RUFAS/routines/feed_storage/hay.py:28]: Should soy be here? Somewhere in the discussion of setting of crop quality defaults, we came to the conclusion that soy should also be able to be harvested as hay, and there is currently a soy hay option for in the available crops. If this is a mistake we should remove the soy hay option from the Crop module.

File: RUFAS/routines/feed_storage/storage.py [hunk 6]
```diff
@@ -35,24 +42,58 @@ class Storage:
...
-        pass
+
+        Raises
+        ------
+            NotImplementedError
+                If the storage's acceptable crops is not populated.
+            ValueError
+                If the crop's category is not compatible with the storage.
+            Exception
+                If adding the crop exceeds the storage's capacity.
```
Comment (Isgotkowitz) [inline RUFAS/routines/feed_storage/storage.py]:
```suggestion
        Raises
        ------
        NotImplementedError
            If the storage's acceptable crops is not populated.
        ValueError
            If the crop's category is not compatible with the storage.
        Exception
            If adding the crop exceeds the storage's capacity.
```

**Review by Isgotkowitz — CHANGES_REQUESTED** [review]
Changes requested on docstring formatting. I am not sure exactly how docstrings for Enums should be formatted (the Numpy docstring guide does not provide any explicit guidance), but I think generally they should look a lot like regular class docstrings. Had some other comments about which will help make sure that the Crop module will line up with the Feed Storage module correctly.

File: RUFAS/routines/feed_storage/enums.py (hunk 3 above)
Comment (PooyaHekmati) [reply RUFAS/routines/feed_storage/enums.py]: That is a question for the SME @KevinP-B .

File: RUFAS/routines/feed_storage/baleage.py (hunk 4 above)
Comment (PooyaHekmati) [reply RUFAS/routines/feed_storage/baleage.py:26]: It is already documented in the parent class.

File: RUFAS/routines/feed_storage/hay.py (hunk 5 above)
Comment (PooyaHekmati) [reply RUFAS/routines/feed_storage/hay.py:28]: That is a question for SME @KevinP-B .

File: RUFAS/routines/feed_storage/storage.py [hunk 7]
```diff
@@ -35,24 +42,58 @@ class Storage:
...
+        if not self.acceptable_crops:
+            raise NotImplementedError(
+                "Storage.acceptable_crops is not populated, consider populating it in the child class."
+            )
+        if crop.category not in self.acceptable_crops:
+            raise ValueError(
+                f"Can't recieve the crop, the compatible crop categories are {self.acceptable_crops=},\
+                    {crop.category} is not one of them."
+            )
+        if self.stored_mass + crop.fresh_mass > self.capacity:
```
Comment (Isgotkowitz) [inline RUFAS/routines/feed_storage/storage.py:89]:
```suggestion
        unacceptable_crop = crop.category not in self.acceptable_crops
        if unacceptable_crop:
            raise ValueError(
                f"Can't recieve the crop, the compatible crop categories are {self.acceptable_crops=},\
                    {crop.category} is not one of them."
            )
        capacity_exceeded = self.stored_mass + crop.fresh_mass > self.capacity
        if capacity_exceeded:
```
Missed these the first time through. Up to you if they are strictly necessary though.

**Review by Isgotkowitz — CHANGES_REQUESTED** [review]
Changes requested on the attributes section of the docstring of `CropCategory`. Also found some if conditions that could be broken down a bit, not sure how necessary those changes are though.

File: RUFAS/routines/feed_storage/storage.py (hunk 7 above)
Comment (PooyaHekmati) [reply RUFAS/routines/feed_storage/storage.py:89]: I have mentioned this in some other PRs before: we use documentary variables when understanding the statement is difficult. Here, the next line is explaining what is happening. Putting every single statement in variables is not a good practice.

**Review by Isgotkowitz — APPROVED** [review]
LG8TM, nice job Pooya.

**Review by KevinP-B — APPROVED** [review]
Efficient updates and ensuring crops match appropriate storage options elegantly

**Pull request merged**